src/platform/Darwin/BleConnectionDelegateImpl.mm

```diff
@@ -283,20 +275,49 @@ - (BOOL)isConnecting
 - (void)resetTimer
 {
     if ([self isConnecting]) {
-        auto timeout = static_cast<int64_t>(kScanningWithDiscriminatorTimeoutInSeconds * NSEC_PER_SEC);
-        dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, timeout), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
+        [self updateTimer:kScanningWithDiscriminatorTimeoutInSeconds];
     } else if ([self isScanningWithoutDelegate]) {
-        auto timeout = static_cast<int64_t>(kScanningWithoutDelegateTimeoutInSeconds * NSEC_PER_SEC);
-        dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, timeout), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
+        [self updateTimer:kScanningWithoutDelegateTimeoutInSeconds];
     } else if ([self isScanning]) {
-        dispatch_source_cancel(_timer);
+        [self clearTimer];
     } else {
         // It should not happens.
         [self stop];
         [self dispatchConnectionError:CHIP_ERROR_INCORRECT_STATE];
     }
 }
 
+- (void)setupTimer
+{
+    if (_timer) {
+        [self clearTimer];
+    }
+
+    _timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, _workQueue);
+    dispatch_source_set_event_handler(_timer, ^{
+        [self stop];
+        [self dispatchConnectionError:BLE_ERROR_APP_CLOSED_CONNECTION];
+    });
+    [self resetTimer];
+    dispatch_resume(_timer);
```

Comment:
I am lost in this logic.  Who is expected to `setupTimer`?  Under what conditions?  `setupTimer` is allowed to set `_timer` to nil, in which case how is this `dispatch_resume` ok?

Reply:
I have updated the code. Let me know if that's clearer.